# Week 2 Lab — Classes & Objects (OOP - I)

**192-201 Advanced Computer Programming with Generative AI**

Reference: *Think Python, 3rd Edition* — Chapter 14 (Classes & Functions) + Chapter 15 (Classes & Methods).
Practice problems adapted from *Python 100 Programs* (Devbrat Rudra), **SET-2 #10** and **SET-10 #8/#9**.

---

### About this lab

This is a **guided practice notebook** — it is **not graded**. Its job is to help you write **classes**: bundles that keep data and behaviour together, so you can create your own kinds of objects.

Work through the parts **in order**. For each part:
1. Read the short guidance.
2. **Run every code cell** and look at the result.
3. Complete each **Your turn** / **TODO** task.

> **Run a cell:** click it, then press **Shift + Enter**.

> **AI-use policy — Level 1 (explain only):** you may ask an AI assistant to *explain* an idea, but **type all the code yourself**. Do not paste AI-written code today.

## Part 0 — Warm-up: from a function to an object

Previouly you organised code with **functions**. This week we bundle data **and** the functions that work on it into a single **object**. First, run this reminder of the old way:

In [ ]:
# The 192-101 way: data and function are separate
balance = 100

def deposit(current, amount):
    return current + amount

balance = deposit(balance, 50)
print(balance)   # 150

150


> Nothing links `balance` and `deposit` together — you must pass the balance in every time. A **class** fixes this by keeping them in one place. That is what we build today.

## Part 1 — Define a class and make an object

A **class** is a blueprint for a kind of object. Use the `class` keyword and a `CapWords` or  `Pascal Case` name. Run this tiny class:

In [ ]:
class Dog:
    """A dog with a name."""

    def __init__(self, name):   # runs automatically when you make a Dog
        self.name = name        # store the name on THIS object
d1 = Dog('Rex')                 # d1 is an object (an instance) of Dog
print(d1.name)                  # Rex

Rex


Key words to notice:
- **class** `Dog` — the blueprint.
- **object / instance** `d1` — a real dog built from the blueprint.
- **attribute** `name` — data stored on the object, read with `d1.name`.
- **`__init__`** — the special method that runs automatically to set up a new object.
- **`self`** — the object being built or used (here, `d1`).

**Your turn.** Make a class called `Cat` with a `name` attribute (copy the pattern above). Then create a cat called `'Milo'` and print its name.

In [ ]:
class Cat:
    """A cat with a name."""

    def __init__(self, name):
        self.name = name

    def poop(self, name):
        print(f'{name} pooped')

    def eat(self, name):
        print(f'{name} ate')

c1 = Cat('Milo')
print(c1.name)
c1.poop('Milo')
c1.eat('Milo')


Milo
Milo pooped
Milo ate


## Part 2 — Objects can hold several attributes

An object usually stores more than one piece of data. Pass several values into `__init__` and store each with `self.`. This is the **Basic Class and Object** example from your workbook *(Python 100 Programs, SET-2 #10)*.

In [ ]:
class Rectangle:
    """A rectangle with a length and a width."""

    def __init__(self, length, width):
        self.length = length    # attribute 1
        self.width = width      # attribute 2

r = Rectangle(4, 5)
print(r.length)   # 4
print(r.width)    # 5

**Your turn.** Make a class `Student` with **two** attributes set in `__init__`: `name` and `student_id`. Create a student `'Mai'` with id `6501`, then print both attributes.

In [ ]:
class Student:
    """A student with a name and student id."""

    def __init__(self, name, student_id):
        self.name = name
        self.student_id = student_id

name = input('Enter name: ')
student_id = input('Enter student id: ')
s1 = Student(name, student_id)
print(s1.name)
print(s1.student_id)


Enter name: Lwin Oo
Enter student id: 67890
Lwin Oo
67890


## Part 3 — Methods: behaviour that uses the object's data

A **method** is a function defined *inside* a class. It always takes `self` first, so it can use the object's own attributes. Here we add an `area()` method to `Rectangle` — completing the workbook example:

In [ ]:
class Rectangle:
    """A rectangle with a length, a width, and an area."""

    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):                     # a method (behaviour)
        return self.length * self.width

r = Rectangle(4, 5)
print(r.area())   # 20  <- called ON the object with r.area()

20


Notice `r.area()` is called **on the object**. Inside the method, `self` *is* `r`, so `self.length` is `4` and `self.width` is `5`.

**Your turn.** Add a method `perimeter(self)` to the `Rectangle` class that returns `2 * (length + width)`. Then create a rectangle and print its perimeter.

*Tip:* copy the class above into the cell, then add your new method next to `area`.

In [ ]:

class Rectangle:
    """A rectangle with a length, a width, and an area."""

    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):                     # a method (behaviour)
        return self.length * self.width

    def perimeter(self):
        return 2 * (self.length * self.width)

r = Rectangle(4, 5)
print(f'area : {r.area()}')   # 20  <- called ON the object with r.area()
print(f'perimeter : {r.perimeter()}' )

area : 20
perimeter : 40


## Part 4 — Make your object readable with `__str__`

By default, printing an object shows something unhelpful like `<__main__.Rectangle object at 0x...>`. Run this to see the problem:

In [1]:
r = Rectangle(4, 5)
print(r)     # unhelpful memory address

NameError: name 'Rectangle' is not defined

Add a special method `__str__` that **returns** a readable string. Python calls it automatically whenever you `print` the object:

In [2]:
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return self.length * self.width

    def __str__(self):                  # decides how the object looks as text
        return f'Rectangle {self.length}x{self.width} = {self.area()}'

r = Rectangle(4, 5)
print(r)     # Rectangle 4x5 = 20

Rectangle 4x5 = 20


> Remember: `__str__` must **return** a string (not `print` it). Then `print(r)` and `str(r)` both use it.

**Your turn.** Add a `__str__` to your `Student` class from Part 2 so that `print(student)` shows: `Student: Mai (6501)`.

In [4]:
class Student:
    """A student with a name and student id."""

    def __init__(self, name, student_id):
        self.name = name
        self.student_id = student_id

    def __str__(self):
      return f'Student: {self.name} ({self.student_id})'

student = Student('Mai', 6501)
print(student)


Student: Mai (6501)


## Part 5 — Guided build: a `BankAccount` class

Now put it all together — attributes, methods, and `__str__` — in one class. This mirrors the **Counter** class in your workbook *(Python 100 Programs, SET-10 #8)*, where methods change an attribute.

Fill in the `# TODO` lines. The first method is done for you as an example.

In [5]:
class BankAccount:
    """A simple bank account."""

    def __init__(self, owner, balance=0):
        self.owner = owner
        self.balance = balance

    def deposit(self, amount):          # example — done for you
        self.balance = self.balance + amount

    def withdraw(self, amount):
        # TODO: subtract amount from self.balance
        self.balance = self.balance - amount

    def __str__(self):
        # TODO: return f'{self.owner}: ${self.balance}'
        return f'{self.owner}: ${self.balance}'

acct = BankAccount('Mai', 100)
acct.deposit(50)
acct.withdraw(30)      # <- uncomment after you finish withdraw
print(acct.balance)      # 150 for now
print(acct)            # <- uncomment after you finish __str__

120
Mai: $120


**Check your work.** After you complete `withdraw` and `__str__`, uncomment the two lines. `acct.deposit(50)` then `acct.withdraw(30)` on a start of `100` should leave `120`, and `print(acct)` should show `Mai: $120`.

## Part 6 — Build it yourself: a `Student` with grades

From this short spec, write a `Student` class **from scratch** (this is the core Week 2 skill). Type it yourself — **AI-use Level 1**.

**Specification**
1. **Attributes** (set in `__init__`): `name`, `student_id`, and an empty list `grades`.
2. **Method** `add_grade(self, g)` — append `g` to `self.grades`.
3. **Method** `average(self)` — return the mean of `self.grades` (return `0` if the list is empty).
4. **`__str__`** — return e.g. `Mai (6501): avg 3.50`.
5. **Exercise it** — create 2 students, add a few grades to each, and print each one.

In [9]:
class Student:
    """A student with a name, id, and grades."""

    def __init__(self, name, student_id):
        # TODO: set self.name, self.student_id, and self.grades = []
        self.name = name
        self.student_id = student_id
        self.grades = []


    def add_grade(self, g):
        # TODO: append g to self.grades
        self.grades.append(g)


    def average(self):
        # TODO: return the mean of self.grades (0 if empty)
        if len(self.grades) == 0:
            return 0
        else:
            return sum(self.grades) / len(self.grades)


    def __str__(self):
        # TODO: return f'{self.name} ({self.student_id}): avg {self.average():.2f}'
        return f'{self.name} ({self.student_id}): avg {self.average():.2f}'

# TODO: create 2 students, add grades, and print each one
student1 = Student('Mai', 6501)
student2 = Student('John', 6502)
student1.add_grade(3.5)
student1.add_grade(4.0)
student1.add_grade(3.0)
student2.add_grade(3.0)
student2.add_grade(3.5)
student2.add_grade(2.0)

print(student1)
print(student2)


Mai (6501): avg 3.50
John (6502): avg 2.83


## Part 7 — Class vs Object (work in pairs)

With a partner, pick **one** class from this lab and describe it. Separating the **blueprint** from the **objects** is the key idea of OOP.

**Example — Rectangle:**

| Class (blueprint) | Attributes (data) | Methods (behaviour) | Example objects |
|---|---|---|---|
| `Rectangle` | length, width | area, __str__ | `Rectangle(4,5)`, `Rectangle(2,9)` |

**Your turn.** Double-click the table below to edit it, and fill in the class you chose:

| Class (blueprint) | Attributes (data) | Methods (behaviour) | Example objects |
|---|---|---|---|
| *(your class)* | *(its data)* | *(its methods)* | *(two objects)* |

## Part 8 — Learning with an AI assistant (Level 1)

You may ask an assistant to **explain**, not to write your code. Good Level-1 questions:

- "Explain what `self` means in a method, with a simple example."
- "Why does `__init__` run automatically when I create an object?"
- "What is the difference between a class and an object?"
- "I got an `AttributeError`. In general, what causes that?"

**Not allowed today:** asking the assistant to write the `Student` or `BankAccount` class for you and pasting it in.

## Before you finish — quick checklist

You should now be able to:

- [ ] Define a class with `class Name:` and an `__init__` method
- [ ] Store data on an object using `self.attribute = value`
- [ ] Create objects (instances) and read their attributes with dot notation
- [ ] Add **methods** that use `self` to reach the object's data
- [ ] Give a class a readable **`__str__`**
- [ ] Complete the `Rectangle`, `BankAccount`, and `Student` classes

**Before Week 3:** make sure your `Student` class runs correctly and is pushed to your GitHub repo. Next week we add **encapsulation** — using `__init__` and `@property` so an object can never hold invalid data.

*Nicely done.*